QEfficient/utils/_utils.py

```diff
@@ -0,0 +1,124 @@
+
+from QEfficient.utils.constants import QEFF_MODELS_DIR
+
+
+def login_and_download_hf_lm(pretrained_model_name_or_path, *args, **kwargs):
+    hf_token = kwargs.pop("hf_token", None)
+    cache_dir = kwargs.pop("cache_dir", None)   
+    if hf_token is not None:
+        login(hf_token)
+    pretrained_model_name_or_path = hf_download(
+        repo_id=pretrained_model_name_or_path,
+        cache_dir=cache_dir,
+        ignore_patterns=["*.txt", "*.onnx", "*.ot", "*.md", "*.tflite", "*.pdf", "*.msgpack", "*.h5"],
+    )
+    return pretrained_model_name_or_path
+
+
+def hf_download(
+    repo_id: Optional[str] = None,
+    cache_dir: Optional[str] = None,
+    hf_token: Optional[str] = None,
+    allow_patterns: Optional[List[str]] = None,
+    ignore_patterns: Optional[List[str]] = None,
+):
+    # Setup cache and local dir
+    local_dir = None
+    if cache_dir is not None:
+        cache_dir = f"{cache_dir}"
+        local_dir = f"{cache_dir}/{repo_id}"
+
+    os.makedirs(f"{cache_dir}/{repo_id}", exist_ok=True)
+    max_retries = 5
+    retry_count = 0
+    while retry_count < max_retries:
+        try:
+            model_path = snapshot_download(
+                repo_id,
+                cache_dir=cache_dir,
+                local_dir=local_dir,
+                local_dir_use_symlinks=True,
+                revision="main",
+                resume_download=True,
+                token=hf_token,
+                allow_patterns=allow_patterns,
+                ignore_patterns=ignore_patterns,
+            )
+            break
+        except requests.ReadTimeout as e:
+            print(f"Read timeout: {e}")
+            retry_count += 1
+
+        except HTTPError as e:
+            retry_count = max_retries
+            if e.response.status_code == 401:
+                print("You need to pass a valid `--hf_token=...` to download private checkpoints.")
+            else:
+                raise e
+
+    return model_path
+
+
+def qpc_exists(model_name: str, qpc_base_dir_name: str) -> Tuple[bool, str]:
+    """
+    Checks if qpc files already exists, removes the directory if files have been manipulated.
+    ---------
+    :param model_name: str. HF Model card name.
+    :param dir_path: str. Path of qpc directory.
+    :return: Union[Tuple[bool, str]]: qpc_exists and path to qpc directory
+    """
+    model_card_dir = os.path.join(QEFF_MODELS_DIR, str(model_name))
+    os.makedirs(model_card_dir, exist_ok=True)
+
+    qpc_dir_path = os.path.join(model_card_dir, qpc_base_dir_name, "qpcs")
+
+    # Compute the boolean indicating if the QPC exists
+    qpc_exists_bool = os.path.isdir(qpc_dir_path) and os.path.isfile(os.path.join(qpc_dir_path, "programqpc.bin"))
+
+    return qpc_exists_bool, qpc_dir_path
+
+
+def onnx_exists(model_name: str) -> Tuple[bool, str, str]:
+    """
+    Checks if qpc files already exists, removes the directory if files have been manipulated.
```

Comment:
needs correction

Reply:
Updated, missed this as this function was copied from another file. In this PR I was only changing location of this function.